src/goSuggest.ts

```diff
@@ -75,6 +81,44 @@ export class GoCompletionItemProvider implements vscode.CompletionItemProvider {
 		});
 	}
 
+	public resolveCompletionItem(item: vscode.CompletionItem, token: vscode.CancellationToken): vscode.ProviderResult<vscode.CompletionItem> {
+		if (item instanceof ExtendedCompletionItem) {
+			const goRuntimePath = getBinPath('go');
+			if (!goRuntimePath) {
+				vscode.window.showInformationMessage('Cannot find "go" binary. Update PATH or GOROOT appropriately');
+				return;
+			}
+
+			const env = getToolsEnvVars();
+			const cwd = path.dirname(item.document.fileName);
+
+			return new Promise<vscode.CompletionItem>((resolve, reject) => {
+				let args = ['doc', '-c', '-cmd', '-u'];
+				if (item.package) {
+					args.push(item.package, item.label);
+				} else {
+					args.push(item.label);
+				}
+
+				cp.execFile(goRuntimePath, args, { cwd }, (err, stdout, stderr) => {
+					if (err) {
+						reject(err);
```

Comment:
In case of error, we do not want the completion feature to fail. So lets `return resolve(item)` here.

Reply:
How VS Code behaves in regards to this callback failing seems poorly documented. I will have to check this.

Reply:
I think that's the code handling this: [vscode/src/vs/editor/contrib/suggest/suggest widget.ts:609-631](https://github.com/Microsoft/vscode/blob/f13c8d1eab216fb84472a68c3ba8c3d9d76b1a30/src/vs/editor/contrib/suggest/suggestWidget.ts#L609-631).

It looks like it simply carries on without the extra information and doesn't abort the completion. Probably logging the error into one of the output streams. That's how I remember it behaving too, but please verify me. If you feel more comfortable with just `resolve(item)` than I can change it as such. But in that case we should log the error ourself somewhere. Otherwise it would be impossible to debug errors here.

Reply:
https://github.com/Microsoft/vscode/blob/f13c8d1eab216fb84472a68c3ba8c3d9d76b1a30/src/vs/editor/contrib/suggest/suggest.ts#L157 is where the `resolveCompletionItem` gets called. When it doesnt return anything, `value` is `undefined`.

The `assign` function then tries to run `Object.keys()` on it which would give a `Cannot convert undefined or null to object` error.

Error messages can be logged using console.log such that if anyone wants to investigate why docs are not present, the logs can help.